Libro/controller

GuoqingLibro opened the pull request.

GuoqingLibro committed:
Second updating based on the required changes

The required changes based on some conventions I made wrong

GuoqingLibro committed:
Fixing the error by moving the kenel.file

GuoqingLibro committed:
wip

GuoqingLibro committed:
Merge branch 'main' into Libro/controller

GuoqingLibro committed:
Controller for celebrity

controller building up for celebrity
celebrity data can be updated and edited

vrodriguero commented on routes/web.php:
```diff
@@ -17,6 +17,24 @@
+Route::put('celebrities/{celebrity}', [CelebrityController::class, 'update'])->name('celebrities.update');
+
+Route::delete('celebrities/{celebrity}', [CelebrityController::class, 'destroy'])->name('celebrities.destroy');
+
+
+
+
```
you can remove these extra lines here ;D just to make things a bit cleaner.

vrodriguero commented on tests/Feature/CelebrityTest.php:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @return void
+     */
+    public function test_example()
+    {
+        $response = $this->get('/');
+
+        $response->assertStatus(200);
+    }
```
This test is only testing the index page, so because we are testing the celebrities this test can be deleted. 

By convention, we usually keep only the tests that are relevant to the celebrity here.

vrodriguero commented on tests/Feature/CelebrityTest.php:
```diff
@@ -0,0 +1,34 @@
+    /** @test  */
+    public function test_users_can_browse_celebrity()
+    {
+        $user = User::factory()->create();
+
+        $response = $this->get('/celebrities');
+
+        $response -> assertStatus(200);
```
So this can be simplified to something like this:

```
$user = User::factory()->create();

$this->actingAs($user)
       ->get('/celebrities')
       ->assertOk()
```
The example above is a bit more readable don't you think?
- So you create an user
- Acting as the user, you try to get to the `/celebrities` route, then the expected response should be an Ok (status code 200)

vrodriguero commented on tests/Unit/CelebrityTest.php:
```diff
@@ -0,0 +1,18 @@
+
+use PHPUnit\Framework\TestCase;
+
+class CelebrityTest extends TestCase
+{
+    /**
+     * A basic unit test example.
+     *
+     * @return void
+     */
+    public function test_example()
+    {
+        $this->assertTrue(true);
+    }
+}
```
You can delete this file ;D 

Let's focus on the Http(Feature) tests for now

vrodriguero replied:
I think you missed the comment above ;D

vrodriguero commented on composer.json:
```diff
@@ -12,6 +12,7 @@
         "laravel/tinker": "^2.5"
     },
     "require-dev": {
+        "brianium/paratest": "^6.3",
```
How is this package being used at the moment? 😄

vrodriguero commented on app/Http/Controllers/CelebrityController.php:
```diff
@@ -2,9 +2,64 @@
+    public function index() //print all data in Celebrity database
+    {
+        return view('celebrities.index', [
+            'celebrities' => Celebrity::all()
```
For this one Libro, I think we need to be careful when getting all the data like this. This can be really heavy on the DB side and not really user-friendly. 

For example:
What would happen if we had 1 million celebrities? 

We should consider:
Should we show all of the results? 
Should we paginate? 
And what about the order of the records?

vrodriguero reviewed (changes requested):
@GuoqingLibro Good effort,

I have left some suggestions for you ;D and I would also suggest that now that you have the main into your branch, you update your CelebrityMovies factory to use a movie id instead of a fake int. And the celebrity model if needed, to create a relation with a movie :D

GuoqingLibro committed:
Required changes updated

added paginator
fixed some required changes

GuoqingLibro committed:
The follow up updates on layout for celebrity

this is the follow up based on the previous changes which has been done.
the change is about layout for celebrity pages

GuoqingLibro committed:
follow up change for fixing the factory for CelebrityMovie Factory

the CelebrityMovie relation factory now using real data instead of using faker to generate.

vrodriguero commented on app/Http/Controllers/CelebrityController.php:
```diff
@@ -2,9 +2,65 @@
 namespace App\Http\Controllers;
 
+use Carbon\Carbon;
+use Illuminate\Support\Facades\DB;
```
Do we need this DB here? 
If not we can remove it ;D

vrodriguero commented on app/Http/Controllers/CelebrityController.php:
```diff
@@ -2,9 +2,65 @@
+    public function index() //print all data in Celebrity database
+    {
+        return view('celebrities.index', [
+            'celebrities' => Celebrity::paginate(5)
```
Hey @GuoqingLibro good use of paginate. But I think that we should make this a bit more refined. 

When passing something to a view, we should also consider how do we want our data to be displayed. Like for example, the order of our data.

vrodriguero commented on app/Http/Controllers/CelebrityController.php:
```diff
@@ -2,9 +2,65 @@
+    public function show(Celebrity $celebrity)
+    {
+        return view('celebrities.show', compact('celebrity'));
+        // similar to return view('celebrities.show', $celebrity=>celebrity);
```
I think we can remove this comment ;D

vrodriguero commented on app/Http/Controllers/CelebrityController.php:
```diff
@@ -2,9 +2,65 @@
+            'nationality' => ['required', 'string', 'min:4'],
+            'bio' => ['required', 'string'],
+        ]));
+        return redirect()->route('celebrities.show', $celebrity)->with('success', 'Celebrity added');
```
Usually, before the return statement, we have a blank line for code readability. 😄

vrodriguero commented on database/factories/CelebrityMovieFactory.php:
```diff
@@ -23,8 +24,8 @@ class CelebrityMovieFactory extends Factory
-            'celebrity_id' => Celebrity::inRandomOrder()->first()->id,
-            'movie_id' => $this->faker->randomDigitNot(8),
+            'celebrity_id' =>Celebrity::inRandomOrder()->first()->id,
+            'movie_id' =>Movie::inRandomOrder()->first()->id
```
Awesome!

vrodriguero commented on resources/views/celebrities/index.blade.php:
```diff
@@ -0,0 +1,23 @@
+<x-app-layout>
+    <x-slot name="header">
+        Celebrity Page
```
Perhaps we can change that to Celebrities Page?

vrodriguero commented on resources/views/celebrities/show.blade.php:
```diff
@@ -0,0 +1,35 @@
+<x-app-layout>
+    <x-slot name="header">
+        Celebrity Profile
```
Maybe it is worth showing the Celebrity name in the header, instead of Celebrity Profile?

vrodriguero commented on resources/views/celebrities/show.blade.php:
```diff
@@ -0,0 +1,35 @@
+
+            <div>
+                <div>
+                    Bio: {{$celebrity->bio}}
```
Missing a closing div here. Make sure you format your code regularly to avoid issues and bugs ;D 

The VScode shortcut to format a file is:
` Format Document (Ctrl+Shift+I)`

vrodriguero commented on routes/web.php:
```diff
@@ -17,6 +17,20 @@
     return view('welcome');
 });
 
+Route::get('celebrities', [CelebrityController::class, 'index'])->name('celebrities.index');
+
+Route::get('celebrity/create', [CelebrityController::class, 'create'])->name('celebrities.create');
+
+Route::get('celebrities/{celebrity}/edit', [CelebrityController::class, 'edit'])->name('celebrities.edit');
+
+Route::get('celebrities/{celebrity}', [CelebrityController::class, 'show'])->name('celebrities.show');
+
+Route::post('celebrities', [CelebrityController::class, 'store'])->name('celebrities.store');
+
+Route::put('celebrities/{celebrity}', [CelebrityController::class, 'update'])->name('celebrities.update');
+
+Route::delete('celebrities/{celebrity}', [CelebrityController::class, 'destroy'])->name('celebrities.destroy');
+
```
I think that chat a bit yesterday about this ;D But there are some of these routes that should be only seen by authenticated users. 
You can use the `Route::group` and use the middleware `auth` to make some of these routes only accessible to authenticated users.

vrodriguero commented on tests/Feature/CelebrityTest.php:
```diff
@@ -0,0 +1,19 @@
+    {
+        $user = User::factory()->create();
+
+        $this->actingAs($user)->get('/celebrities')->assertOk();
```
Formatting a bit here ;D The best way would be to the get and the assert ok in different lines. 

```
$this->actingAs($user)
       ->get...
        ->assertOk...
```

Also try to test the store and update and validation rules ;D

vrodriguero reviewed (changes requested):
Hi @GuoqingLibro Great effort! 

I hope you are getting a bit more familiar with the code. 

I left some comments and suggestions, but please feel free to contact me if you did not understand something. 😄 

Happy coding 💻

vrodriguero commented on resources/views/celebrities/create.blade.php:
```diff
@@ -0,0 +1,49 @@
+            </div>
+            <div>
+                Bio: 
+                <input
+                    type="text"
+                    name="bio"
+                    id="bio"
+                    value= "{{ old('bio') }}"
```
With longer texts input I would suggest to use a textarea instead of the input. 

Also, try to use the components that are available to you here ;D You can take a look at the login and see how they used the `input` and the `label` components.

vrodriguero reviewed (changes requested):
One more comment 😄

GuoqingLibro committed:
required changes has been fixed

Fixed the required changes
Used some of the available component as the label in the  blade

GuoqingLibro committed:
required updating

the paiginate has been tune up
the teaxtarea has been updated
eadit page also been fixed same as create page by using x-app-layout

vrodriguero commented on tests/Feature/CelebrityTest.php:
```diff
@@ -0,0 +1,20 @@
+    {
+        $user = User::factory()->create();
+
+        $this->actingAs($user)->get('/celebrities')
+                              ->assertOk();
```
Usually when we have a sequence that has `::` or `->`. We format it in a way where each statement after the first one stays in its own line. 
```
$this->actingAs.... 
       ->get()
       ->assert 
```
Something like this.

vrodriguero commented on resources/views/celebrities/edit.blade.php:
```diff
@@ -0,0 +1,36 @@
+            <div>
+                <x-label for="bio" value="Bio"/>
+
+                <textarea type="bio" name="bio"  rows="5" cols="25" id="bio" :value="old('bio', $celebrity->bio)"></textarea>
```
Libro have you tested this manually? You will find that using the value attribute like that will not work with textarea elements. 
So for example, when you want to pass the value to a textarea it should be inside the brackets.

vrodriguero replied:
You can try to look for some examples online too :D

vrodriguero reviewed (changes requested):
Hey @GuoqingLibro 

Just a really small change but you are on the right track! 

😄

GuoqingLibro committed:
middleware updating

middleware for users and admin has been added
the changes required has been fixed

GuoqingLibro committed:
required changes updated and test case generated

the last required changes updated
the test case has been generated for admin/user on all methods

vrodriguero commented on tests/Feature/CelebrityTest.php:
```diff
@@ -44,4 +45,56 @@ public function test_admin_can_access_celebrities_create_page()
+
+        $this->actingAs($user)
+                ->delete('/celebrities/'.$celebrity->id)
+                ->assertRedirect('/');
```
Hey Libro

Whenever we are testing something that might change the DB, we also need to check if the DB is as expected. 

You can do that by using `assertDatabaseHas` or `assertDatabaseMissing` in the case of an unauthorized delete action like this one. We want o check if the record is still on the Database. So you would use the `assertDatabaseHas` to do that. 

If an admin deletes a record, we want to ensure that the record has been soft deleted in the database (that is if the class allows soft deletes) 😄

vrodriguero commented on app/Http/Middleware/IsAdmin.php:
```diff
@@ -0,0 +1,26 @@
+    public function handle($request, Closure $next)
+    {
+        if ( !Auth::check() || Auth::check() && !Auth::user()->is_admin) {
+            return redirect('/')->with('error','You have not admin access');
```
Maybe `You are not authorised to perform this action`.

vrodriguero commented on routes/web.php:
```diff
@@ -17,8 +17,24 @@
+
+Route::get('celebrities/{celebrity}', [CelebrityController::class, 'show'])->name('celebrities.show');
+
+Route::post('celebrities', [CelebrityController::class, 'store'])->name('celebrities.store');
```
This should probably be in the group right?

vrodriguero reviewed (changes requested):
Hey Libro, 

Left some suggestions for you 😄

GuoqingLibro committed:
layout of celebrity page updated

the data with celebrity and related information has been showed up on page
the layout of celebrity has been updated

vrodriguero commented on database/factories/CelebrityFactory.php:
```diff
@@ -26,6 +26,16 @@ public function definition()
+    public function randomposter()
+    {
+        $posters = [
+            'https://upload.wikimedia.org/wikipedia/commons/4/46/Leonardo_Dicaprio_Cannes_2019.jpg'
```
Probably try to look for more pictures to put in here, that way it will look much more dynamic ;D

vrodriguero commented on resources/views/celebrities/create.blade.php:
```diff
@@ -1,34 +1,70 @@
     <x-slot name="header">
-        Create Celebrity  
+        <h2 class="font-semibold text-xl leading-tight">
+            {{ __('Add a new celebrity to Database') }}
```
My suggestion is to keep the public pages without any terms that might be confusing to other people. The database is a term in the IT industry, normal users might not know what that means. 

I would suggest changing this to Create a celebrity or Add a new celebrity...

vrodriguero commented on resources/views/test.blade.php:
```diff
@@ -0,0 +1,33 @@
+{{-- THIS IS A PHP FOR SIMPLE TESTING. DELETE THIS ONCE ALL TESTING IS COMPLETED (OR UPON MERGE WITH Finn'S BRANCH) --}}
```
You can delete this. You do not need a test view for the movie ;D

vrodriguero commented on app/Http/Controllers/CelebrityController.php:
```diff
@@ -17,7 +18,9 @@ public function index() //print all data in Celebrity database
     public function show(Celebrity $celebrity)
     {
-        return view('celebrities.show', compact('celebrity'));
+
```
You can delete this empty line

vrodriguero reviewed (changes requested):
Good @GuoqingLibro !

Now you can try to add the movie to a celebrity when creating and updating ;D

GuoqingLibro committed:
updated required changes and movies are showed up on the create page

the required changes has been fixed
the movies are now appear on the create and edit celebrity page and tested working

vrodriguero committed:
fix layout for celebrities

vrodriguero committed:
Merge branch 'main' into Libro/controller

# Conflicts:
#	app/Http/Controllers/CelebrityController.php
#	app/Models/Celebrity.php
#	database/factories/CelebrityMovieFactory.php
#	database/factories/MovieFactory.php
#	database/factories/UserFactory.php
#	database/seeders/DatabaseSeeder.php
#	resources/views/components/form.blade.php
#	resources/views/layouts/app.blade.php
#	resources/views/layouts/navigation.blade.php
#	routes/web.php

vrodriguero committed:
fix active link

vrodriguero committed:
fix link

vrodriguero committed:
merge main
